src/main/java/jenkins/plugins/mailer/tasks/MimeMessageBuilder.java

```diff
@@ -0,0 +1,264 @@
+        }
+        return this;
+    }
+
+    public MimeMessageBuilder setSubject(String subject) {
+        this.subject = subject;
+        return this;
+    }
+
+    public MimeMessageBuilder setBody(String body) {
+        this.body = body;
+        return this;
+    }
+
+    public MimeMessageBuilder setRecipientFilter(AddressFilter recipientFilter) {
+        this.recipientFilter = recipientFilter;
+        return this;
+    }
+
+    public MimeMessageBuilder addRecipients(String recipients) throws UnsupportedEncodingException {
+        addRecipients(recipients, Message.RecipientType.TO);
+        return this;
+    }
+
+    public MimeMessageBuilder addRecipients(String recipients, Message.RecipientType recipientType) throws UnsupportedEncodingException {
+        if (StringUtils.isNotBlank(recipients) && recipientType != null) {
+            StringTokenizer tokens = new StringTokenizer(recipients);
+            while (tokens.hasMoreTokens()) {
+                String addressToken = tokens.nextToken();
+                InternetAddress internetAddress = toNormalizedAddress(addressToken);
+
+                if (internetAddress != null) {
+                    if (recipientType == Message.RecipientType.TO) {
+                        to.add(internetAddress);
+                    } else if (recipientType == Message.RecipientType.CC) {
+                        cc.add(internetAddress);
+                    } else if (recipientType == Message.RecipientType.BCC) {
+                        bcc.add(internetAddress);
+                    }
+                }
+            }
+        }
+
+        return this;
+    }
+
+    /**
+     * Build a {@link MimeMessage} instance from the set of supplied parameters.
+     * @return The {@link MimeMessage} instance;
+     * @throws MessagingException
+     * @throws UnsupportedEncodingException
+     */
+    public MimeMessage buildMimeMessage() throws MessagingException, UnsupportedEncodingException {
+        MimeMessage msg = new MimeMessage(Mailer.descriptor().createSession());
+
+        setJenkinsInstanceIdent(msg);
+
+        msg.setContent("", mimeType);
+        if (StringUtils.isNotBlank(from)) {
+            msg.setFrom(toNormalizedAddress(from));
+        }
+        msg.setSentDate(new Date());
+
+        addSubject(msg);
+        addBody(msg);
+        addRecipients(msg);
+
+        if (StringUtils.isNotBlank(replyTo)) {
+            msg.setReplyTo(new Address[]{toNormalizedAddress(replyTo)});
+        }
+        return msg;
+    }
+
+    private void setJenkinsInstanceIdent(MimeMessage msg) throws MessagingException {
+        if (Jenkins.getInstance() != null) {
+            RSAPublicKey publicKey = InstanceIdentity.get().getPublic();
+            msg.setHeader("X-Instance-Identity", Base64.encode(publicKey.getEncoded()));
+        }
+    }
+
+    public static void setInReplyTo(MimeMessage msg, String inReplyTo) throws MessagingException {
+        if (msg == null || inReplyTo == null) {
```

Comment:
Again, why is it useful to allow someone to pass null for these parameters? Prefer everything to be `@Nonnull` unless there is a compelling usability reason why clients should be permitted to sometimes pass null (in which case the parameter should be `@CheckForNull` to make this clear).